corfudb-tools/src/main/java/org/corfudb/browser/CorfuOfflineBrowserEditor.java

```diff
@@ -0,0 +1,458 @@
+        return null;
+    }
+
+    @Override
+    public CorfuDynamicRecord editRecord(String namespace, String tableName, String keyToEdit, String newRecord) {
+        return null;
+    }
+
+    @Override
+    public int deleteRecordsFromFile(String namespace, String tableName, String pathToKeysFile, int batchSize) {
+        return 0;
+    }
+
+    @Override
+    public int deleteRecords(String namespace, String tableName, List<String> keysToDelete, int batchSize) {
+        return 0;
+    }
+
+    @Override
+    public int loadTable(String namespace, String tableName, int numItems, int batchSize, int itemSize) {
+        return 0;
+    }
+
+    @Override
+    public int listenOnTable(String namespace, String tableName, int stopAfter) {
+        return 0;
+    }
+
+    @Override
+    public Set<String> listStreamTags() {
+        return null;
+    }
+
+    @Override
+    public Map<String, List<CorfuStoreMetadata.TableName>> listTagToTableMap() {
+        return null;
+    }
+
+    @Override
+    public Set<String> listTagsForTable(String namespace, String table) {
+        return null;
+    }
+
+    @Override
+    public List<CorfuStoreMetadata.TableName> listTablesForTag(@Nonnull String streamTag) {
+        return null;
+    }
+
+    class CorfuTableDescriptor {
+        @Getter
+        private final UUID streamID;
+        @Getter
+        private final UUID checkpointID;
+        private final String namespace;
+        private final String tableName;
+
+        public CorfuTableDescriptor(String namespace, String tableName) {
+            this.namespace = namespace;
+            this.tableName = tableName;
+            String name = TableRegistry.getFullyQualifiedTableName(namespace, tableName);
+            streamID = CorfuRuntime.getStreamID(name);
+            checkpointID = CorfuRuntime.getCheckpointStreamIdFromId(streamID);
+        }
+
+        public boolean belongsToStream(LogData data) {
+            return data.containsStream(streamID) || data.containsStream(checkpointID);
+        }
+
+        public boolean belongsToStream(String namespace, String tableName) {
+            return this.namespace.equals(namespace) && this.tableName.equals(tableName);
+        }
+    }
+
+    class OrderedObject {
+        @Getter
+        private Object update;
+        @Getter
+        private long order;
+
+        public OrderedObject(Object smrObject, long address) {
+            this.update = smrObject;
+            this.order = address;
+        }
+    }
+
+    class CachedTable {
+        private ConcurrentMap table;
+        private long clearTableTxn;
+
+        public CachedTable() {
+            this.table = new ConcurrentHashMap<>();
+            this.clearTableTxn = -1;
+        }
+
+        public void insert(Object key, Object value, long address) {
+            if ((table.containsKey(key)) && (((OrderedObject) table.get(key)).getOrder() > address)) {
```

Comment:
What is the real type of `key`? If it's Object then using containsKey is not enough since it only compares object references. You can check how the online corfu browser uses CorfuDynamicMessage.

Reply:
We are using the same struct to store RegistryTable, ProtoBufTable, and other Tables data and these tables have keys of classes `TableName`, `ProtobufFileName`, and `CorfuDynamicKey` respectively.
That's the reason, I've used a generic `Object` Type. But I get your point, `hashcode()` for the Object class is returning the reference.
In the Online browser, we use the runtime to get the Table as an ICorfuTable object, but over here we do build data from logs.